WordPress/gutenberg
# Font face permissions

## What?

Modifies the font face permissions to ensure the file system is writable for the installation and deletion of fonts.

The font library's JavaScript is modified to use these permission to determine whether to allow:

* installation of google fonts
* upload of font files
* deletion of fonts

See #55280

## Why?

Currently the `DISALLOW_FILE_MODS` constant is not checked as part of the font faces API allowing users to upload fonts when the fonts should not be disabled.

As this constant is used to indicate an immutable file system, the constant needs to be considered to allow for a good user experience uploading fonts. See https://github.com/WordPress/wporg-main-2022/issues/388, https://github.com/WordPress/gutenberg/issues/55063

## How?

* Induces the meta capabilities: `install_font_faces`, `delete_font_faces`.
* Checks `[core data store].canUser( 'create', 'font-families' );` to determine if a user can upload, removes the UI if not
* Checks `[core data store].canUser( 'delete', 'font-families' );` to determine if a user can delete, removes the UI if not

> [!NOTE]
>
> Please test running this branch on **WordPress 6.4**.
>
> The file the changes have been made in are not loaded on the current version of WordPress trunk.


## Testing Instructions

1. Install a font either via file upload or Google Fonts.
2. Add `define( 'DISALLOW_FILE_MODS', true );` to your wp-config file.
3. Open Site Editor ... Styles > Typography > Manage Fonts
4. Ensure the upload and Install fonts tabs are not shown
5. Click the font you installed earlier
6. Check the box next to the font
7. Ensure the delete button is not available
8. Remove the constant added earlier
9. Ensure it's possible to delete your font.

### Testing Instructions for Keyboard


## Screenshots or screencast
